### PR TITLE
Generator APIs: get ready for validation

### PR DESCRIPTION
**Pull request by vlada-shubina**
## `Abstractions`
- improved `IGenerator` APIs 
- exposed validation errors in `ITemplate`

## `Orchestrator.RunnableProjects`
To get ready for validation, the correct entry point for it is needed. `TemplateConfigModel` is not sufficient, as it doesn't know about template sources. `RunnableProjectConfig` is already overloaded.

I created an abstract intermediate layer for it: `DirectoryBasedTemplate`. This class will also implement validation further on.
There ate 2 implementations of `DirectoryBasedTemplate`
- `ScannedTemplateInfo` - lightweight implementation containing info about all localization available. It is used as the result of template scanning but it is not ready to be run.
- `RunnableProjectConfig` - remains entry point for running the template. 

`RunnableProjectConfig` kept test constructor accepting `TemplateConfigModel`. This option should not be used for production implementation as it is not fully ready for `ITemplate` or `ITemplateInfo` (because it doesn't have `IFile` for configuration).

`ITemplateValidationInfo` is preliminary interface to be exposed to validation component. It will remain in `Orchestartor.RunnableProjects`. The only interface to be put to `Abstractions` is `IValidationEntry`.

----
Remains unsolved:
`ITemplate` is now used in two scenarios:
- scanning via `GetTemplatesFromMountPoint`
- instantiating template via `TryLoadTemplateFromTemplateInfo`

These usages are different for localization:
- on scanning all localizations are loaded and validated
- on instantiation only relevant localization is loaded

It would be better to split them but decided to keep untouched for now. `Localizations` won't be populated in instantiation scenario; `ValidationErrors` won't have validation issues on non-relevant localization files on instantiation.

Potential `IGenerator` API alternative A: 
```csharp
//unchanged
Task<ICreationResult> CreateAsync(
            IEngineEnvironmentSettings environmentSettings,
            ITemplate template,
            IParameterSetData parameters,
            string targetDirectory,
            CancellationToken cancellationToken);

//unchanged
Task<ICreationEffects> GetCreationEffectsAsync(
            IEngineEnvironmentSettings environmentSettings,
            ITemplate template,
            IParameterSetData parameters,
            string targetDirectory,
            CancellationToken cancellationToken);
            
//ITemplateInfo (without `LocaleConfigPlace`) + locs
IReadOnlyList<ITemplateWithLocInfo> GetTemplatesFromMountPoint(IMountPoint source); 

//ITemplateLocator: MountPointUri, ConfigPlace, LocaleConfigPlace, HostConfigPlace
bool TryLoadTemplateFromTemplateInfo(IEngineEnvironmentSettings settings, ITemplateLocator config, out ITemplate? template, string? baselineName = null);
```

Potential `IGenerator` API alternative B: 
```csharp
//unchanged
Task<ICreationResult> CreateAsync(
            IEngineEnvironmentSettings environmentSettings,
            ITemplate template,
            IParameterSetData parameters,
            string targetDirectory,
            CancellationToken cancellationToken);

//unchanged
Task<ICreationEffects> GetCreationEffectsAsync(
            IEngineEnvironmentSettings environmentSettings,
            ITemplate template,
            IParameterSetData parameters,
            string targetDirectory,
            CancellationToken cancellationToken);
            
//ITemplateInfo - already returns correct locale
IReadOnlyList<ITemplateInfo> GetTemplatesFromMountPoint(IMountPoint source, string locale); 

//unchanged; or may be switched to ITemplateLocator.
bool TryLoadTemplateFromTemplateInfo(IEngineEnvironmentSettings settings, ITemplateInfo config, out ITemplate? template, string? baselineName = null);
```
----

Small refactoring:
- removed `ISimpleConfigModifiers` and its implementation and replaced it with `string`
- split former `SpecialConfigParams` to `OperationConfigDefault`.

### Checks:
- [ ] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Abstractions/IValidationEntry.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        /// <summary>
+        /// Error severity.
+        /// </summary>
+        public enum SeverityLevel
+        {
+            None,
+            Error,
+            Warning,
+            Info
+        }
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Abstractions/IValidationEntry.cs:20]: Unusual ordering

**Review by JanKrivanek — COMMENTED** [review]
Looks like work in progress - I'll have few more passes through the changes

**Comment by vlada-shubina** [comment]
Rebased it and now ready for review. I don't plan to add more logic. 
Switch to validation components will happen in next PR as it gets big.

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/DirectoryBasedTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            }
+        }
+
+        /// <summary>
+        /// The method validates loaded cpnfiguration. The errors and warnings are printed using logger.
+        /// The warning messages is mainly only for template authoring, and should be used for template validation, so for now they are logged in debug level only.
+        /// https://github.com/dotnet/templating/issues/2623.
+        /// </summary>
+        /// <exception cref="TemplateValidationException">in case of validation fails.</exception>
+        private void PerformTemplateValidation()
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/DirectoryBasedTemplate.cs:308]: Should this actually return list of `IValidationEntry`? Similarly with other checking methods in this template type

**Review by JanKrivanek — APPROVED** [review]
Overal the restrucutalizations makes sense and simplify some monster-classes.

The existing logic is still kept as fast-fail, unstructuralized validation - nut I gues this is planned to be areworked at next steps

**Comment by vlada-shubina** [comment]
I'll open new PR for this

**Pull request closed without merging**